src/conv-core/cpuaffinity.C

```diff
@@ -442,62 +448,149 @@ static void cpuAffinityHandler(void *m)
+        if (rankmsg->node_hosts[i] == rankmsg->node_hosts[j] &&
+              rankmsg->node_ranks[i] > rankmsg->node_ranks[j])
+          std::swap(rankmsg->node_ranks[i], rankmsg->node_ranks[j]);
+      }
     for (i=0; i<npes-1; i++)
       for(j=i+1; j<npes; j++) {
-        if (rankmsg->nodes[i] == rankmsg->nodes[j] && 
-              rankmsg->ranks[i] > rankmsg->ranks[j]) 
-          std::swap(rankmsg->ranks[i], rankmsg->ranks[j]);
+        if (rankmsg->pe_hosts[i] == rankmsg->pe_hosts[j] &&
+              rankmsg->pe_ranks[i] > rankmsg->pe_ranks[j])
+          std::swap(rankmsg->pe_ranks[i], rankmsg->pe_ranks[j]);
       }
     }
 #endif
-    CmiSyncBroadcastAllAndFree(sizeof(rankMsg)+CmiNumPes()*sizeof(int)*2, (void *)rankmsg);
+    CmiSyncBroadcastAllAndFree(RANKMSG_SIZE(nnodes, npes), (void *)rankmsg);
+  }
+}
+
+static inline std::vector<hwloc_obj_t> getPUListForAutoAffinity(hwloc_obj_t container, unsigned int nodesize)
+{
+  std::vector<hwloc_obj_t> pu_list;
+  std::queue<std::pair<hwloc_obj_t, unsigned int>> thisqueue, nextqueue;
+
+  nextqueue.emplace(container, 0);
+
+  do
+  {
+    std::swap(thisqueue, nextqueue);
+
+    do
+    {
+      auto entry = thisqueue.front();
+      thisqueue.pop();
+
+      if (entry.first->type == HWLOC_OBJ_PU && !in_exclude(entry.first->os_index))
+      {
+        pu_list.push_back(entry.first);
+
+        if (pu_list.size() == nodesize)
+          return pu_list;
+      }
+
+      if (entry.second < entry.first->arity)
+      {
+        nextqueue.emplace(entry.first->children[entry.second++], 0);
+
+        if (entry.second < entry.first->arity)
+          thisqueue.emplace(entry);
+      }
+    }
+    while (!thisqueue.empty());
   }
+  while (!nextqueue.empty());
+
+  return pu_list;
 }
 
 /* called on each processor */
 static void cpuAffinityRecvHandler(void *msg)
 {
-  int myrank, mynode;
+  const int mynode = CmiMyNode();
+  const int mype = CmiMyPe();
+  const int numnodes = CmiNumNodes();
+  const int numpes = CmiNumPes();
+
   rankMsg *m = (rankMsg *)msg;
-  m->ranks = (int *)((char*)m + sizeof(rankMsg));
-  m->nodes = (int *)((char*)m + sizeof(rankMsg) + CmiNumPes()*sizeof(int));
-  myrank = m->ranks[CmiMyPe()];
-  mynode = m->nodes[CmiMyPe()];
+  size_t siz = sizeof(rankMsg);
+  m->node_ranks = (int *)((char*)m + siz);
+  siz += CmiNumNodes()*sizeof(int);
+  m->node_hosts = (int *)((char*)m + siz);
+  siz += CmiNumNodes()*sizeof(int);
+  m->pe_ranks = (int *)((char*)m + siz);
+  siz += CmiNumPes()*sizeof(int);
+  m->pe_hosts = (int *)((char*)m + siz);
+  siz += CmiNumPes()*sizeof(int);
+  CmiAssert(siz == RANKMSG_SIZE(numnodes, numpes));
+
+  int node_rank = m->node_ranks[mynode];
+  int pe_rank = m->pe_ranks[mype];
+  int myhost = m->pe_hosts[mype];
+
+  int nodesOnHost = 0;
+  for (int i = 0; i < numnodes; ++i)
+    if (m->node_hosts[i] == myhost)
+      ++nodesOnHost;
+
+  hwloc_obj_t obj;
+  const int package_count = cmi_hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_PACKAGE);
```

Comment:
Some comments on what this part of the code is doing would be nice.

Reply:
Added.